go/arrow/compute/exec.go

```diff
@@ -0,0 +1,163 @@
+// limitations under the License.
+
+package compute
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/apache/arrow/go/v10/arrow"
+	"github.com/apache/arrow/go/v10/arrow/compute/internal/exec"
+	"github.com/apache/arrow/go/v10/arrow/internal/debug"
+)
+
+func haveChunkedArray(values []Datum) bool {
+	for _, v := range values {
+		if v.Kind() == KindChunked {
+			return true
+		}
+	}
+	return false
+}
+
+// ExecSpanFromBatch constructs and returns a new ExecSpan from the values
+// inside of the ExecBatch which could be scalar or arrays.
+//
+// This is mostly used for tests but is also a convenience method for other
+// cases.
+func ExecSpanFromBatch(batch *ExecBatch) *exec.ExecSpan {
+	out := &exec.ExecSpan{Len: batch.Len, Values: make([]exec.ExecValue, len(batch.Values))}
+	for i, v := range batch.Values {
+		outVal := &out.Values[i]
+		if v.Kind() == KindScalar {
+			outVal.Scalar = v.(*ScalarDatum).Value
+		} else {
+			outVal.Array.SetMembers(v.(*ArrayDatum).Value)
+			outVal.Scalar = nil
+		}
+	}
+	return out
+}
+
+// this is the primary driver of execution
+func execInternal(ctx context.Context, fn Function, opts FunctionOptions, passedLen int64, args ...Datum) (result Datum, err error) {
+	if opts == nil {
+		if err = checkOptions(fn, opts); err != nil {
+			return
+		}
+		opts = fn.DefaultOptions()
+	}
+
+	// we only allow Array, ChunkedArray, and Scalars for now.
+	// RecordBatch and Table datums are disallowed.
+	if err = checkAllIsValue(args); err != nil {
+		return
+	}
+
+	inTypes := make([]arrow.DataType, len(args))
+	for i, a := range args {
+		inTypes[i] = a.(ArrayLikeDatum).Type()
+	}
+
+	var (
+		k        exec.Kernel
+		executor kernelExecutor
+	)
+
+	switch fn.Kind() {
+	case FuncScalar:
+		executor = scalarExecPool.Get().(*scalarExecutor)
+		defer func() {
+			executor.clear()
+			scalarExecPool.Put(executor.(*scalarExecutor))
+		}()
+	default:
+		return nil, fmt.Errorf("%w: direct execution of %s", arrow.ErrNotImplemented, fn.Kind())
+	}
+
+	if k, err = fn.DispatchBest(inTypes...); err != nil {
+		return
+	}
+
+	kctx := &exec.KernelCtx{Ctx: ctx, Kernel: k}
+	init := k.GetInitFn()
+	kinitArgs := exec.KernelInitArgs{Kernel: k, Inputs: inTypes, Options: opts}
+	if init != nil {
+		kctx.State, err = init(kctx, kinitArgs)
+		if err != nil {
+			return
+		}
+	}
+
+	if err = executor.Init(kctx, kinitArgs); err != nil {
+		return
+	}
+
+	input := ExecBatch{Values: args, Len: 0}
+	if input.NumValues() == 0 {
+		if passedLen != -1 {
+			input.Len = passedLen
+		}
+	} else {
+		inferred, _ := inferBatchLength(input.Values)
+		input.Len = inferred
+		switch fn.Kind() {
+		case FuncScalar:
+			if passedLen != -1 && passedLen != inferred {
+				return nil, fmt.Errorf("%w: passed batch length for execution did not match actual length for scalar fn execution",
+					arrow.ErrInvalid)
+			}
+		}
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	ch := make(chan Datum, 10)
```

Comment:
`10` looks without strong support, but I don't have better suggestion.

Is channel strictly necessary here? I think channel make the code easier to read than C++ which collects all results with a datum accumulator.

Reply:
You're correct that a channel is not strictly necessary here, I did it for the reason you suggested: it makes it more readable than the C++ using a datum accumulator.

I'll make the `10` a configurable setting via the `ExecCtx` and just leave 10 as the default for now. Ultimately i just wanted to make sure that neither the `Exec` nor the `WrapResults` got blocked in most cases on how quickly the execution works.